### PR TITLE
Avoid const return type in functions

### PR DESCRIPTION
**Pull request by masterleinad**
`const` doesn't do anything when we are returning values.

### CODE REVIEW

File: include/deal.II/algorithms/any_data.h [hunk 1]
```diff
@@ -76,7 +76,7 @@ class AnyData : public Subscriptor
    * use try_read() instead.
    */
   template <typename type>
-  const type
+  type
   entry(const std::string &name) const;
```
Comment (bangerth) [inline include/deal.II/algorithms/any_data.h:80]: Could `type` be a reference/pointer? And if so, would that make the pointer const, or the pointed to object const?

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
Several of the testers fail, at least one of them with this error message:
```
/home/runner/work/dealii/dealii/source/base/mpi.cc:681:5: error: ambiguating new declaration of ‘const std::vector<unsigned int> dealii::Utilities::MPI::mpi_processes_within_communicator(MPI_Comm, MPI_Comm)’
  681 |     mpi_processes_within_communicator(const MPI_Comm, const MPI_Comm)
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/runner/work/dealii/dealii/source/base/mpi.cc:19:
/home/runner/work/dealii/dealii/include/deal.II/base/mpi.h:158:5: note: old declaration ‘std::vector<unsigned int> dealii::Utilities::MPI::mpi_processes_within_communicator(MPI_Comm, MPI_Comm)’
  158 |     mpi_processes_within_communicator(const MPI_Comm comm_large,
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: include/deal.II/algorithms/any_data.h (hunk 1 above)
Comment (masterleinad) [reply include/deal.II/algorithms/any_data.h:80]: The `const` doesn't matter in that case either, see https://godbolt.org/z/8jesMca3a. Returning a constant pointer is the same as returning a non-constant pointer and a pointer to a constant type is still returned as such.

**Comment by bangerth** [comment]
Still fails:
```
/jenkins/workspace/dealii-serial_PR-15671/tests/multigrid/mg_data_out_01.cc: In instantiation of ‘void print(dealii::DataOut<dim>&, const dealii::Triangulation<dim, dim>&) [with int dim = 2]’:
/jenkins/workspace/dealii-serial_PR-15671/tests/multigrid/mg_data_out_01.cc:62:8:   required from ‘void do_test() [with int dim = 2]’
/jenkins/workspace/dealii-serial_PR-15671/tests/multigrid/mg_data_out_01.cc:95:14:   required from here
/jenkins/workspace/dealii-serial_PR-15671/tests/multigrid/mg_data_out_01.cc:34:9: error: cannot bind non-const lvalue reference of type ‘std::pair<std::function<dealii::TriaIterator<dealii::CellAccessor<2, 2> >(const dealii::Triangulation<2, 2>&)>, std::function<dealii::TriaIterator<dealii::CellAccessor<2, 2> >(const dealii::Triangulation<2, 2>&, const dealii::TriaIterator<dealii::CellAccessor<2, 2> >&)> >&’ to an rvalue of type ‘std::pair<std::function<dealii::TriaIterator<dealii::CellAccessor<2, 2> >(const dealii::Triangulation<2, 2>&)>, std::function<dealii::TriaIterator<dealii::CellAccessor<2, 2> >(const dealii::Triangulation<2, 2>&, const dealii::TriaIterator<dealii::CellAccessor<2, 2> >&)> >’
   34 |   auto &p = data_out.get_cell_selection();
      |         ^
/jenkins/workspace/dealii-serial_PR-15671/tests/multigrid/mg_data_out_01.cc: In instantiation of ‘void print(dealii::DataOut<dim>&, const dealii::Triangulation<dim, dim>&) [with int dim = 3]’:
/jenkins/workspace/dealii-serial_PR-15671/tests/multigrid/mg_data_out_01.cc:62:8:   required from ‘void do_test() [with int dim = 3]’
/jenkins/workspace/dealii-serial_PR-15671/tests/multigrid/mg_data_out_01.cc:96:14:   required from here
/jenkins/workspace/dealii-serial_PR-15671/tests/multigrid/mg_data_out_01.cc:34:9: error: cannot bind non-const lvalue reference of type ‘std::pair<std::function<dealii::TriaIterator<dealii::CellAccessor<3, 3> >(const dealii::Triangulation<3, 3>&)>, std::function<dealii::TriaIterator<dealii::CellAccessor<3, 3> >(const dealii::Triangulation<3, 3>&, const dealii::TriaIterator<dealii::CellAccessor<3, 3> >&)> >&’ to an rvalue of type ‘std::pair<std::function<dealii::TriaIterator<dealii::CellAccessor<3, 3> >(const dealii::Triangulation<3, 3>&)>, std::function<dealii::TriaIterator<dealii::CellAccessor<3, 3> >(const dealii::Triangulation<3, 3>&, const dealii::TriaIterator<dealii::CellAccessor<3, 3> >&)> >’
make[3]: *** [CMakeFiles/multigrid.mg_data_out_01.debug.dir/build.make:76: CMakeFiles/multigrid.mg_data_out_01.debug.dir/mg_data_out_01.cc.o] Error 1
make[2]: *** [CMakeFiles/Makefile2:1902: CMakeFiles/multigrid.mg_data_out_01.debug.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:2143: CMakeFiles/multigrid.mg_data_out_01.debug.test.dir/rule] Error 2
```

**Comment by bangerth** [comment]
That is a pretty fascinating failure, by the way. It shows that the `const` on an rvalue return object really *does* matter in some contexts!

**Comment by bangerth** [comment]
In short, this does not compile:
```
const int f();

int &i = f();
```
The error in the test probably points to a case where we shouldn't store a reference to begin with.

**Comment by masterleinad** [comment]
> That is a pretty fascinating failure, by the way. It shows that the `const` on an rvalue return object really _does_ matter in some contexts!

Yes, also see https://godbolt.org/z/r5nhc6Yde. `auto&` becomes `const Foo&` when the return type is `const`.

**Comment by bangerth** [comment]
I wonder whether we want to reconsider this. It breaks ASPECT in ~4 places: https://github.com/geodynamics/aspect/pull/5208

**Pull request merged**